### PR TITLE
Manage kubevirt CR from sdk generated code

### PR DESCRIPTION
**Pull request by rthallisey**
# Hyperconverged Cluster Operator

The goal of this operator is to provide a single entrypoint for multiple operators - kubevirt, cdi, networking, ect... - where users can deploy and configure them in a single object.  This operator is sometimes referred to as a "meta operator" or an "operator for operators".  Most importantly, this operator doesn't replace or interfere with OLM.  It only creates operator CRs, which is the user's prerogative.

This patch:
  - Generate the hco code from operator-sdk
  - make some changes to create kubevirt's CR when an hco CR is created

### CODE REVIEW

**Comment by booxter** [comment]
@rthallisey how do I deploy it on a ready cluster? I see a bunch of manifests under `deploy` directory. Should I apply all of them or not? If all of them, is there any order of application expected, or it's idempotent? I think this should be explained in a README file.

I think you would want to remove `build/_output/` directory from the PR since it's artifacts. (Perhaps also update `.gitignore` to skip such files.)

**Comment by booxter** [comment]
Perhaps I am doing it wrong but when I applied all the manifests from the `deploy` directory (first those under `crd`, then `role*.yaml`, then `service_account.yaml` and finally the operator manifest) I get the operator pod failing as follows:

```
{"level":"error","ts":1552427075.3559277,"logger":"kubebuilder.source","msg":"if kind is a CRD, it should be installed before calling Start","kind":"KubeVirt.kubevirt.io","error":"no matches for kind \"KubeVirt\" in version \"kubevirt.io/v1alpha3\"","stacktrace":"github.com/kubevirt/hyperconverged-cluster-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/vendor/github.com/go-logr/zapr/zapr.go:128\ngithub.com/kubevirt/hyperconverged-cluster-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source.(*Kind).Start\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/vendor/sigs.k8s.io/controller-runtime/pkg/source/source.go:89\ngithub.com/kubevirt/hyperconverged-cluster-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Watch\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:122\ngithub.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/hyperconverged.add\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/hyperconverged/hyperconverged_controller.go:52\ngithub.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/hyperconverged.Add\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/hyperconverged/hyperconverged_controller.go:28\ngithub.com/kubevirt/hyperconverged-cluster-operator/pkg/controller.AddToManager\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/pkg/controller/controller.go:13\nmain.main\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/cmd/manager/main.go:103\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:201"}
{"level":"error","ts":1552427075.3560224,"logger":"cmd","msg":"","error":"no matches for kind \"KubeVirt\" in version \"kubevirt.io/v1alpha3\"","stacktrace":"github.com/kubevirt/hyperconverged-cluster-operator/vendor/github.com/go-logr/zapr.(*zapLogger).Error\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/vendor/github.com/go-logr/zapr/zapr.go:128\nmain.main\n\t/home/rhallisey/src/github.com/kubevirt/hyperconverged-cluster-operator/cmd/manager/main.go:104\nruntime.main\n\t/usr/lib/golang/src/runtime/proc.go:201"}
```

It looks like it complains about `KubeVirt` CRD not created. I thought it's the job for the operator to pull in all the relevant pieces and create them in the right order, no?

**Comment by rthallisey** [comment]
@booxter added a README.  This is only the first PR towards this operator so it's in a very early phase, but we need more PRs :).

> I think you would want to remove build/_output/
>
Should be replaced by a Makefile, but it will come in another commit.

> Perhaps I am doing it wrong but when I applied all the manifests from the deploy directory (first those under crd, then role*.yaml, then service_account.yaml and finally the operator manifest) I get the operator pod failing as follows:
>
You're running it correctly, since the HCO expects the kubevirt CRD to be present.  The kubevirt-operator is responsible for the CRD, so you get the CRD added to your cluster when your launch that operator with `oc create -f https://github.com/kubevirt/kubevirt/releases/download/v0.15.0/kubevirt-operator.yaml`

File: deploy/crds/hco_v1alpha1_hyperconverged_crd.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
+apiVersion: apiextensions.k8s.io/v1beta1
+kind: CustomResourceDefinition
+metadata:
+  name: hyperconvergeds.hco.kubevirt.io
+spec:
+  group: hco.kubevirt.io
```
Comment (djzager) [inline deploy/crds/hco_v1alpha1_hyperconverged_crd.yaml]: Maybe the group should be just `kubevirt.io` or is there some requirement that these be `hyperconverged.hco.kubevirt.io`?

**Comment by rthallisey** [comment]
> One comment and is it possible for you to add the generated code as it's own commit 
>
I'll see what I can do.

File: deploy/crds/hco_v1alpha1_hyperconverged_crd.yaml (hunk 1 above)
Comment (rthallisey) [reply deploy/crds/hco_v1alpha1_hyperconverged_crd.yaml]: Take that back, rbac rules is where there needs to be significance.  Having the same group will make the hco share rules with all kubevirt apis.  Not sure we want to do that just yet.  

I don't think there's harm in leaving as is, so we can revisit later.

**Pull request merged**